src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -85,30 +89,15 @@ public Void run() throws Exception {
      * Gets the next ordinal and throw {@link AbortException} the milestone lives inside a parallel step branch.
      */
     private synchronized int processOrdinal() throws AbortException {
-        // TODO: use FlowNodeSerialWalker when released (if possible)
-        FlowGraphWalker walker = new FlowGraphWalker();
-        walker.addHead(node);
-        Integer previousOrdinal = null;
-        int parallelDetectionEnabled = 0;
-        for (FlowNode n : walker) {
-
-            if (parallelDetectionEnabled <= 0 && n.getAction(ThreadNameAction.class) != null) {
-                throw new AbortException("Using a milestone step inside parallel is not allowed");
-            }
-
-            if (n instanceof BlockEndNode) {
-                parallelDetectionEnabled++;
-            } else if (n instanceof BlockStartNode && !(n instanceof FlowStartNode)) {
-                parallelDetectionEnabled--;
-            }
-
-            OrdinalAction a = n.getAction(OrdinalAction.class);
-            if (a != null) {
-                previousOrdinal = a.ordinal;
-                break;
-            }
+        List<FlowNode> heads = node.getExecution().getCurrentHeads();
+        if (heads.size() > 1) {  // TA-DA!  We're inside a parallel, which is forbidden.
```

Comment:
It is not immediately obvious to me that this is correct. What if the other branch has completed? I suppose the current heads would include this step plus the other branch’s block end node?